sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java

```diff
@@ -82,34 +89,67 @@
 public class CoderRegistry implements CoderProvider {
 
   private static final Logger LOG = LoggerFactory.getLogger(CoderRegistry.class);
+  private static final Map<Class<?>, CoderFactory> REGISTERED_CODER_FACTORIES_PER_CLASS;
+
+  static {
+    Set<CoderRegistrar> registrars = Sets.newTreeSet(ObjectsClassComparator.INSTANCE);
+    registrars.addAll(Lists.newArrayList(
+        ServiceLoader.load(CoderRegistrar.class, ReflectHelpers.findClassLoader())));
```

Comment:
The case when a new Service would be available within the same classloader that loaded the CoderRegistry is unlikely to ever occur. Advanced users which dynamically load more classes into the same classloader can re-register if needed.

I would also rather have the same static mapping shared until all places which use a CoderRegistry get it from the Pipeline instead of creating it themselves (e.g. StateSpec).